lib/embulk/input/jira_api/client.rb

```diff
@@ -1,37 +1,75 @@
 require "jiralicious"
 require "parallel"
+require "limiter"
 require "embulk/input/jira_api/issue"
 require "timeout"
 
 module Embulk
   module Input
     module JiraApi
       class Client
-        PARALLEL_THREAD_COUNT = 50
-        SEARCH_TIMEOUT_SECONDS = 5
-        SEARCH_ISSUES_TIMEOUT_SECONDS = 60
+        MAX_RATE_LIMIT = 50
+        MIN_RATE_LIMIT = 2
+        # Normal http request timeout is 300s
+        SEARCH_ISSUES_TIMEOUT_SECONDS = 300
         DEFAULT_SEARCH_RETRY_TIMES = 10
 
+        def initialize
+          @rate_limiter = Limiter::RateQueue.new(MAX_RATE_LIMIT, interval: 2)
+        end
+
         def self.setup(&block)
           Jiralicious.configure(&block)
           new
         end
 
         def search_issues(jql, options={})
-          timeout_and_retry(SEARCH_ISSUES_TIMEOUT_SECONDS) do
-            issues_raw = search(jql, options).issues_raw
-
-            # TODO: below code has race-conditon.
-            Parallel.map(issues_raw, in_threads: PARALLEL_THREAD_COUNT) do |issue_raw|
-              # https://github.com/dorack/jiralicious/blob/v0.4.0/lib/jiralicious/search_result.rb#L32-34
-              issue = Jiralicious::Issue.find(issue_raw["key"])
-              JiraApi::Issue.new(issue)
+          issues_raw = search(jql, options).issues_raw
+          # Maximum number of issues to retrieve is 50
+          rate_limit = MAX_RATE_LIMIT
+          success_items = []
+          fail_items = []
+          error_object = nil
+          timeout_and_retry(SEARCH_ISSUES_TIMEOUT_SECONDS * MAX_RATE_LIMIT ) do
+            retry_count = 0
+            semaphore = Mutex.new
+            @rate_limiter = Limiter::RateQueue.new(rate_limit, interval: 2)
+            error_object = nil
+            while issues_raw.length > 0 && retry_count <= DEFAULT_SEARCH_RETRY_TIMES do
+              search_results = Parallel.each(issues_raw, in_threads: rate_limit) do |issue_raw|
+                # https://github.com/dorack/jiralicious/blob/v0.4.0/lib/jiralicious/search_result.rb#L32-34
+                begin
+                  issue = find_issue(issue_raw["key"])
+                  semaphore.synchronize {
+                    success_items.push(JiraApi::Issue.new(issue))
+                  }
+                rescue MultiJson::ParseError => e
+                  html = e.message
+                  title = html[%r|<title>(.*?)</title>|, 1]
```

Comment:
nit: perhaps using a simple search is more efficient, ie. `html.contains? '<title>Unauthorized (401)</title>'`? I just don't like it when I see RE against a long string such as HTML.